Bug 1266229 - More fixes and tweaks prior to turning on Pulse ingestion

camd opened the pull request:
This contains several tweaks and fixes that allow us to ingest data from
a real Task Cluster owned exchange.

One of the main fixes is the way I was binding to the exchange and
routing keys with the same Queue.  Before, it was re-creating the Queue,
so would miss some of the bindings.

This PR stops just short of actually turning ingestion on.  I figured I'd do that in the next PR to add it to the ``procfile`` and the queue in ``settings.py``

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/mozilla/treeherder/1431)
<!-- Reviewable:end -->

camd commented on treeherder/etl/pulse_consumer.py:
```diff
@@ -16,36 +20,68 @@ class JobConsumer(ConsumerMixin):
+        self.queue.unbind_from(exchange, routing_key)
 
     def on_message(self, body, message):
         try:
```
I had been logging the exchange and routing key on every message here so we could see where messages were coming from.  But I think that'd be way too heavy for real use.  But curious if you had an opinion on some degree of "heartbeat" logging.

edmorley commented:
Reviewed 2 of 4 files at r1.
Review status: 2 of 4 files reviewed at latest revision, 9 unresolved discussions.

---

*[treeherder/config/settings.py, line 181 \[r2\]](https://reviewable.io:443/reviews/mozilla/treeherder/1431#-KIXu0kW1UDg9ze0zvzi:-KIXu0kW1UDg9ze0zvzj:-1834064469) ([raw file](https://github.com/mozilla/treeherder/blob/3b788af151362c6d19efb9db416ca1b184770fd3/treeherder/config/settings.py#L181)):*
> ```Python
>         'kombu': {
>             'handlers': ['console'],
>             'level': 'WARNING',
> ```

Good idea :-)

---

*[treeherder/config/settings.py, line 444 \[r2\]](https://reviewable.io:443/reviews/mozilla/treeherder/1431#-KIXtV87p4eoPozfjL56:-KIXtV87p4eoPozfjL57:-843069808) ([raw file](https://github.com/mozilla/treeherder/blob/3b788af151362c6d19efb9db416ca1b184770fd3/treeherder/config/settings.py#L444)):*
> ```Python
> # Used for making API calls to pulse, such as detecting bindings on the current
> # ingestion queue.
> PULSE_API_URL = env.url("PULSE_API_URL", default="https://pulse.mozilla.org/")
> ```

For this and some of the other settings (eg `PULSE_DATA_INGESTION_QUEUES_DURABLE`), we don't seem to be overriding them anywhere (eg for tests). Unless there's local Vagrant use-cases for them, should we leave them as hardcoded in settings.py rather than pulled from environment variables, until such time as we're actively using them?

---

*[treeherder/config/settings.py, line 453 \[r2\]](https://reviewable.io:443/reviews/mozilla/treeherder/1431#-KIXwI4bjlcWHbPNt51E:-KIXwI4cDJi8d3xca3_k:-1373600147) ([raw file](https://github.com/mozilla/treeherder/blob/3b788af151362c6d19efb9db416ca1b184770fd3/treeherder/config/settings.py#L453)):*
> ```Python
> 
> # Whether to use SSL for the connection for pulse ingestion
> PULSE_DATA_INGESTION_QUEUES_SSL = env.bool("PULSE_DATA_INGESTION_QUEUES_SSL",
> ```

The docs say that the URL can be used to set options, eg `?ssl=1`:
http://docs.celeryproject.org/projects/kombu/en/latest/userguide/connections.html#urls

...so this option could be combined into `PULSE_DATA_INGESTION_CONFIG`?

The docs also aren't clear on what happens if True is passed vs the full dict:
http://docs.celeryproject.org/projects/kombu/en/latest/reference/kombu.connection.html#connection

...ie: does it still verifying the certs or not? Don't suppose you could check? (Would be good to avoid the potential for MITM :-))

---

*[treeherder/etl/pulse_consumer.py, line 25 \[r2\]](https://reviewable.io:443/reviews/mozilla/treeherder/1431#-KIXt53O27UnMWS-pTvy:-KIXt53O27UnMWS-pTvz:1242379138) ([raw file](https://github.com/mozilla/treeherder/blob/3b788af151362c6d19efb9db416ca1b184770fd3/treeherder/etl/pulse_consumer.py#L25)):*
> ```Python
>         self.queue = None
>         config = settings.PULSE_DATA_INGESTION_CONFIG
>         self.queue_name = "queue/{}/jobs".format(urlparse(config).username)
> ```

The default value for `PULSE_DATA_INGESTION_CONFIG` is the empty string. Should `self.queue_name` be instead set to `None`in this case?

---

*[treeherder/etl/pulse_consumer.py, line 54 \[r2\]](https://reviewable.io:443/reviews/mozilla/treeherder/1431#-KIXrqcps4C1VK2k54_4:-KIXrqcps4C1VK2k54_5:-440360459) ([raw file](https://github.com/mozilla/treeherder/blob/3b788af151362c6d19efb9db416ca1b184770fd3/treeherder/etl/pulse_consumer.py#L54)):*
> ```Python
>     def on_message(self, body, message):
>         try:
>             if isinstance(body, dict):
> ```

Is there a reason we're supported both? (It would be hard to remove this later, once people start relying on the flexibility)

---

*[treeherder/etl/pulse_consumer.py, line 66 \[r2\]](https://reviewable.io:443/reviews/mozilla/treeherder/1431#-KIXqsQHC9rd9l7Eg3Wt:-KIXqsQIIzWYWxmJFusj:-393139380) ([raw file](https://github.com/mozilla/treeherder/blob/3b788af151362c6d19efb9db416ca1b184770fd3/treeherder/etl/pulse_consumer.py#L66)):*
> ```Python
>         except:
>             logger.error("Unable to load jobs: {}".format(body), exc_info=1)
>             newrelic.agent.record_exception(params=body)
> ```

What would be the pros cons of replacing `.record_exception()`with re-raising, and putting the `message.ack()` in a `finally` block? (The params for New Relic could be added earlier in this function using `.add_custom_parameter()`)

---

*[treeherder/etl/pulse_consumer.py, line 77 \[r2\]](https://reviewable.io:443/reviews/mozilla/treeherder/1431#-KIXqLMeCcAcj_I3M_8G:-KIXqLMeCcAcj_I3M_8H:1282940465) ([raw file](https://github.com/mozilla/treeherder/blob/3b788af151362c6d19efb9db416ca1b184770fd3/treeherder/etl/pulse_consumer.py#L77)):*
> ```Python
> 
> 
> class PulseApi(JsonExtractorMixin):
> ```

I have a branch locally that tries to remove the last usages of `JsonExtractorMixin`. Could this use the newer `fetch_json()` instead? (it uses requests rather than urllib2)
https://github.com/mozilla/treeherder/blob/9c1b0bf19e6143d4c94a6fcf8efaa7b2b1d7f586/treeherder/etl/common.py#L64

---

*[treeherder/etl/management/commands/ingest_from_pulse.py, line 32 \[r2\]](https://reviewable.io:443/reviews/mozilla/treeherder/1431#-KIXq5ntDi2ffbQMGHnk:-KIXq5nuI52T8_xDyKbl:68849446) ([raw file](https://github.com/mozilla/treeherder/blob/3b788af151362c6d19efb9db416ca1b184770fd3/treeherder/etl/management/commands/ingest_from_pulse.py#L32)):*
> ```Python
>                     "ERROR: Unable to fetch existing bindings for {}".format(
>                         consumer.queue_name))
>                 self.stdout.write("ERROR: Data ingestion may proceed, but no bindings will be pruned")
> ```

Perhaps these should be using stderr instead?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/treeherder/1431)*
<!-- Sent from Reviewable.io -->

edmorley commented:
Reviewed 1 of 4 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/treeherder/1431)*
<!-- Sent from Reviewable.io -->

camd commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[treeherder/config/settings.py, line 444 \[r2\]](https://reviewable.io:443/reviews/mozilla/treeherder/1431#-KIXtV87p4eoPozfjL56:-KIZcOUFEp-dHrHbElEL:-499146009) ([raw file](https://github.com/mozilla/treeherder/blob/3b788af151362c6d19efb9db416ca1b184770fd3/treeherder/config/settings.py#L444)):*
<details><summary><i>Previously, edmorley (Ed Morley) wrote…</i></summary>
> For this and some of the other settings (eg `PULSE_DATA_INGESTION_QUEUES_DURABLE`), we don't seem to be overriding them anywhere (eg for tests). Unless there's local Vagrant use-cases for them, should we leave them as hardcoded in settings.py rather than pulled from environment variables, until such time as we're actively using them?
</details>
Done.

---

*[treeherder/config/settings.py, line 453 \[r2\]](https://reviewable.io:443/reviews/mozilla/treeherder/1431#-KIXwI4bjlcWHbPNt51E:-KIZbpfKoE3rGVicD_9-:1120992364) ([raw file](https://github.com/mozilla/treeherder/blob/3b788af151362c6d19efb9db416ca1b184770fd3/treeherder/config/settings.py#L453)):*
<details><summary><i>Previously, edmorley (Ed Morley) wrote…</i></summary>
> The docs say that the URL can be used to set options, eg `?ssl=1`:
> http://docs.celeryproject.org/projects/kombu/en/latest/userguide/connections.html#urls
> 
> ...so this option could be combined into `PULSE_DATA_INGESTION_CONFIG`?
> 
> The docs also aren't clear on what happens if True is passed vs the full dict:
> http://docs.celeryproject.org/projects/kombu/en/latest/reference/kombu.connection.html#connection
> 
> ...ie: does it still verifying the certs or not? Don't suppose you could check? (Would be good to avoid the potential for MITM :-))
</details>
oh cool.  that's nice wrt the ssl settings.  I modified that.  I must admit, I'm not sure how to check if it's verifying certs or not.  I'll try googling around.

---

*[treeherder/etl/pulse_consumer.py, line 54 \[r2\]](https://reviewable.io:443/reviews/mozilla/treeherder/1431#-KIXrqcps4C1VK2k54_4:-KIZ0FTgZJVPE6VzNnDK:1624168126) ([raw file](https://github.com/mozilla/treeherder/blob/3b788af151362c6d19efb9db416ca1b184770fd3/treeherder/etl/pulse_consumer.py#L54)):*
<details><summary><i>Previously, edmorley (Ed Morley) wrote…</i></summary>
> Is there a reason we're supported both? (It would be hard to remove this later, once people start relying on the flexibility)
</details>
As per vidyo chat, only supporting ``dict`` bodies here.  

---

*[treeherder/etl/pulse_consumer.py, line 66 \[r2\]](https://reviewable.io:443/reviews/mozilla/treeherder/1431#-KIXqsQHC9rd9l7Eg3Wt:-KIZ0NnJdOXunBhy3a45:1078678954) ([raw file](https://github.com/mozilla/treeherder/blob/3b788af151362c6d19efb9db416ca1b184770fd3/treeherder/etl/pulse_consumer.py#L66)):*
<details><summary><i>Previously, edmorley (Ed Morley) wrote…</i></summary>
> What would be the pros cons of replacing `.record_exception()`with re-raising, and putting the `message.ack()` in a `finally` block? (The params for New Relic could be added earlier in this function using `.add_custom_parameter()`)
</details>
Removed exception handling by simplifying the code per vidyo conversation with you.  Any exception will happen in the async task and it will bubble up from there.

---

*[treeherder/etl/pulse_consumer.py, line 77 \[r2\]](https://reviewable.io:443/reviews/mozilla/treeherder/1431#-KIXqLMeCcAcj_I3M_8G:-KIYlLUfrx7OF_ACfDln:-478033148) ([raw file](https://github.com/mozilla/treeherder/blob/3b788af151362c6d19efb9db416ca1b184770fd3/treeherder/etl/pulse_consumer.py#L77)):*
<details><summary><i>Previously, edmorley (Ed Morley) wrote…</i></summary>
> I have a branch locally that tries to remove the last usages of `JsonExtractorMixin`. Could this use the newer `fetch_json()` instead? (it uses requests rather than urllib2)
> https://github.com/mozilla/treeherder/blob/9c1b0bf19e6143d4c94a6fcf8efaa7b2b1d7f586/treeherder/etl/common.py#L64
</details>
oh nice.  yeah, that's perfect.  I'll use that.

---

*[treeherder/etl/management/commands/ingest_from_pulse.py, line 32 \[r2\]](https://reviewable.io:443/reviews/mozilla/treeherder/1431#-KIXq5ntDi2ffbQMGHnk:-KIYjxG2BVCQ7jal41n4:-1475918357) ([raw file](https://github.com/mozilla/treeherder/blob/3b788af151362c6d19efb9db416ca1b184770fd3/treeherder/etl/management/commands/ingest_from_pulse.py#L32)):*
<details><summary><i>Previously, edmorley (Ed Morley) wrote…</i></summary>
> Perhaps these should be using stderr instead?
</details>
ahh, of course.  Yes.  Thanks... :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/treeherder/1431)*
<!-- Sent from Reviewable.io -->

camd commented:
Review status: 1 of 5 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[treeherder/etl/pulse_consumer.py, line 25 \[r2\]](https://reviewable.io:443/reviews/mozilla/treeherder/1431#-KIXt53O27UnMWS-pTvy:-KIZctMg8Qj70utRBhqY:494049150) ([raw file](https://github.com/mozilla/treeherder/blob/3b788af151362c6d19efb9db416ca1b184770fd3/treeherder/etl/pulse_consumer.py#L25)):*
<details><summary><i>Previously, edmorley (Ed Morley) wrote…</i></summary>
> The default value for `PULSE_DATA_INGESTION_CONFIG` is the empty string. Should `self.queue_name` be instead set to `None`in this case?
</details>
Without some value for ``PULSE_DATA_INGESTION_CONFIG`` then all bets are off and nothing will work.  Perhaps I should check it and raise an exception if it's not set, when you try to run the ingestion.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/treeherder/1431)*
<!-- Sent from Reviewable.io -->
